librz/util/float/float.c

```diff
@@ -0,0 +1,1288 @@
+	rz_return_val_if_fail(f, RZ_FLOAT_SPEC_NOT);
+
+	RzFloatSpec ret = RZ_FLOAT_SPEC_NOT;
+	RzBitVector *exp_squashed = get_exp_squashed(f->s, f->r);
+	RzBitVector *mantissa_squashed = get_man_squashed(f->s, f->r);
+	bool sign = get_sign(f->s, f->r);
+
+	if (rz_bv_is_full_vector(exp_squashed)) {
+		/// full exp with 0 mantissa -> inf
+		if (rz_bv_is_zero_vector(mantissa_squashed)) {
+			ret = sign ? RZ_FLOAT_SPEC_PINF : RZ_FLOAT_SPEC_NINF;
+		} else {
+			/// detect signal or quiet nan
+			bool is_quiet = rz_bv_msb(mantissa_squashed);
+			ret = is_quiet ? RZ_FLOAT_SPEC_QNAN : RZ_FLOAT_SPEC_SNAN;
+		}
+	}
+
+	if (rz_bv_is_zero_vector(exp_squashed)) {
+		if (rz_bv_is_zero_vector(mantissa_squashed))
+			ret = RZ_FLOAT_SPEC_ZERO;
+	}
+
+	rz_bv_free(exp_squashed);
+	rz_bv_free(mantissa_squashed);
+
+	return ret;
+}
+
+RZ_API bool rz_float_is_inf(RZ_NONNULL RzFloat *f) {
+	RzFloatSpec type = rz_float_detect_spec(f);
+	if ((type == RZ_FLOAT_SPEC_PINF) || (type == RZ_FLOAT_SPEC_NINF))
+		return true;
+	return false;
+}
+
+RZ_API bool rz_float_is_nan(RZ_NONNULL RzFloat *f) {
+	RzFloatSpec type = rz_float_detect_spec(f);
+	if ((type == RZ_FLOAT_SPEC_SNAN) || (type == RZ_FLOAT_SPEC_QNAN))
+		return true;
+	return false;
+}
+
+static RZ_OWN RzFloat *gen_inf_float(RzFloatFormat format, bool sign) {
+	/// gen a quiet NaN for return
+	RzFloat *ret = rz_float_new(format);
+	RzBitVector *bv = ret->s;
+	ut32 exp_start = rz_float_get_format_info(format, RZ_FLOAT_INFO_MAN_LEN);
+	ut32 exp_end = exp_start + rz_float_get_format_info(format, RZ_FLOAT_INFO_EXP_LEN);
+	for (ut32 i = exp_start; i < exp_end; ++i) {
+		rz_bv_set(bv, i, true);
+	}
+
+	/// set sign bit (MSB)
+	rz_bv_set(bv, bv->len - 1, sign);
+
+	return ret;
+}
+
+static RZ_OWN RzFloat *gen_zero_float(RzFloatFormat format) {
+	RzFloat *ret = rz_float_new(format);
+	rz_float_init(ret);
+	return ret;
+}
+
+static RZ_OWN RzFloat *gen_qnan_float(RzFloatFormat format) {
+	/// gen a quiet NaN for return
+	RzFloat *ret = rz_float_new(format);
+	RzBitVector *bv = ret->s;
+	ut32 exp_start = rz_float_get_format_info(format, RZ_FLOAT_INFO_MAN_LEN);
+	ut32 exp_end = exp_start + rz_float_get_format_info(format, RZ_FLOAT_INFO_EXP_LEN);
+	for (ut32 i = exp_start; i < exp_end; ++i) {
+		rz_bv_set(bv, i, true);
+	}
+	/// set is_quiet to 1
+	rz_bv_set(bv, exp_start - 1, true);
+
+	return ret;
+}
+
+static RZ_OWN RzFloat *propagate_float_nan(RZ_NONNULL RzFloat *left, RzFloatSpec ltype, RZ_NONNULL RzFloat *right, RzFloatSpec rtype) {
+	bool l_is_sig_nan = ltype == RZ_FLOAT_SPEC_SNAN;
+	bool r_is_sig_nan = rtype == RZ_FLOAT_SPEC_SNAN;
+
+	/// gen a quiet NaN for return
+	RzFloatFormat format = left->r;
+	RzFloat *ret = rz_float_new(left->r);
+	RzBitVector *bv = ret->s;
+	ut32 exp_start = rz_float_get_format_info(format, RZ_FLOAT_INFO_MAN_LEN);
+	ut32 exp_end = exp_start + rz_float_get_format_info(format, RZ_FLOAT_INFO_EXP_LEN);
+	for (ut32 i = exp_start; i < exp_end; ++i) {
+		rz_bv_set(bv, i, true);
+	}
```

Comment:
```c
ut32 exp_start = rz_float_get_format_info(format, RZ_FLOAT_INFO_MAN_LEN);
ut32 exp_end = exp_start + rz_float_get_format_info(format, RZ_FLOAT_INFO_EXP_LEN);
for (ut32 i = exp_start; i < exp_end; ++i) {
      rz_bv_set(bv, i, true);
}
```
Since this "set all exp bits" code appears at least three times it could be put in a inline function IMHO.

Reply:
Agree. Thanks

Reply:
I checked the implementations again. they all required the extracted attribute `exp_start` to set flag bit. and the this code snippet is seldom used in other parts. so I think we don't need an inline here. 
maybe an inline function such as `rz_bv_set_range(bitvector, start, end, val)` would be better IMHO

Reply:
Yeah sure. Wasn't so sure about the inline anyways. But just replacing the for loop with some kind of `rz_bv_set_range` already looks better to me.